drivers: power_domain: Power domain TISCI driver support

Kronosblaster opened the pull request:
Support added for power domain regulation using TISCI added for devices using the binding ti,sci-pm-domain.
This driver relies on the TISCI layer to make calls to the device manager core to perform power management.
Power domain support added to AM243x EVM
This PR waits on https://github.com/zephyrproject-rtos/zephyr/pull/90053

## Testing
Test added at tests/power/tisci
Refer: [PowerDomainTest](https://github.com/Kronosblaster/zephyr/blob/tisciLayerTest/samples/drivers/tisci/power_domain/src/main.c)

Output:
```
Starting NULL Bootloader ...                                                                        
                                                                                                    
DMSC Firmware Version 11.0.7--v11.00.07 (Fancy Rat)                                                 
DMSC Firmware revision 0xb                                                                          
DMSC ABI revision 4.0                                                                               
                                                                                                    
INFO: Bootloader_runCpu:205: CPU r5f1-0  is initialized to 800000000 Hz !!!                         
INFO: Bootloader_runCpu:205: CPU r5f1-1 is initialized to 800000000 Hz !!!                          
INFO: Bootloader_runCpu:205: CPU m4f0-0 is initialized to 400000000 Hz !!!                          
INFO: Bootloader_runCpu:205: CPU a530-0 is initialized to 800000000 Hz !!!                          
INFO: Bootloader_runCpu:205: CPU a530-1 is initialized to 800000000 Hz !!!                          
INFO: Bootloader_loadSelfCpu:257: CPU r5f0-0 is initialized to 800000000 Hz !!!                     
INFO: Bootloader_loadSelfCpu:257: CPU r5f0-1 is initialized to 800000000 Hz !!!                     
INFO: Bootloader_runSelfCpu:267: All done, reseting self ...                                        
                                                                                                    
*** Booting Zephyr OS build v4.1.0-5562-g2a492bd9491f ***                                           
Power domain powered OFF (test passed)                                                              
Power domain powered ON (test passed)                                                               
Power domain powered OFF again (test passed)                                                        
All power domain driver tests passed!  
```

Kronosblaster committed:
drivers: power_domain: Power domain TISCI driver support

Support added for power domain regulation using TISCI added for
devices using the binding ti,sci-pm-domain.
This driver relies on the TISCI layer to make calls to the
device manager core to perform power management.

Signed-off-by: Dave Joseph <d-joseph@ti.com>

bjarki-andreasen commented on drivers/power_domain/power_domain_tisci.c:
```diff
@@ -0,0 +1,97 @@
+	int ret;
+
+	switch (action) {
+	case PM_DEVICE_ACTION_RESUME:
+		return 0;
+	case PM_DEVICE_ACTION_SUSPEND:
+		return 0;
```
`PM_DEVICE_ACTION_RESUME` is called when the power domain should power its child devices, `PM_DEVICE_ACTION_SUSPEND` when it should stop powering its child devices. `PM_DEVICE_ACTION_TURN_ON` is called when this power domains power domain is resumed, `PM_DEVICE_ACTION_TURN_OFF` when power domains power domain is suspended

bjarki-andreasen commented on drivers/power_domain/power_domain_tisci.c:
```diff
@@ -0,0 +1,97 @@
+		LOG_ERR("DMSC device not found");
+		return -ENODEV;
+	}
+	return pm_device_runtime_enable(dev);
```
pm_device_driver_init(dev, tisci_pd_pm_action); should be called here, runtime must, for now, be enabled using the devicetree prop `zephyr,pm-device-runtime-auto`, which causes `pm_device_runtime_enable()` to be called by `pm_device_driver_init()`

bjarki-andreasen commented on drivers/power_domain/power_domain_tisci.c:
```diff
@@ -0,0 +1,97 @@
+
+static int tisci_pd_init(const struct device *dev)
+{
+	dmsc = DEVICE_DT_GET(DT_NODELABEL(dmsc));
+	if (dmsc == NULL) {
+		LOG_ERR("DMSC device not found");
+		return -ENODEV;
+	}
```
This does not work, DEVICE_DT_GET() will expand to a token like `__device_dt_ord_20` regardless of whether `dmsc` exists and has a driver enabled. This can be moved to `const struct device *dmsc = DEVICE_DT_GET(DT_NODELABEL(dmsc));` at line 17, or if you want this behavior, DEVICE_DT_GET_OR_NULL() can be used, which expands to NULL if the node dmsc has status = "disabled";

bjarki-andreasen commented on dts/bindings/power-domain/ti,sci-pm-domain.yaml:
```diff
@@ -0,0 +1,22 @@
+include: base.yaml
+
+properties:
+  tisci_device_id:
```
devicetree props use dashes in names: `tisci-device-id`

Kronosblaster commented on dts/bindings/power-domain/ti,sci-pm-domain.yaml:
```diff
@@ -0,0 +1,22 @@
+    description: |
+      The device ID of the power domain as defined in the TISCI documentation.
+  tisci_device_mode:
+    type: int
```
I will use an int based enum. I got include/zephyr/dt-bindings/power/ti,sci_pm_domain.h with the defines anyway. We can have the devicetree do the parsing like you said this way too.

bjarki-andreasen commented on drivers/power_domain/power_domain_tisci.c:
```diff
@@ -0,0 +1,105 @@
+#include <zephyr/logging/log.h>
+#include <zephyr/device.h>
+#include <zephyr/drivers/firmware/tisci/tisci.h>
+#include <zephyr/drivers/power/ti,sci_pm_domain.h>
```
The enum in this file no longer need to be public, the header can be removed and the enums defined in this source file :)

Kronosblaster commented:
> Code looks good :) Do we have a board in tree which uses the power domain we can add the power domain to and test in CI?

We have [AM2434](https://github.com/zephyrproject-rtos/zephyr/pull/87321) but it needs a few modifications to support running tests on it. I can write a test for the driver though that is based off of this board.

Kronosblaster committed:
tests: power: tisci: TISCI power domains test

Test for power domains driver based on TISCI API added. This test is
for the target am243x_evm/am2434/r5f0_0.

Signed-off-by: Dave Joseph <d-joseph@ti.com>

Kronosblaster committed:
dts: arm: ti: PD support for AM243x EVM in DT

Adds power domain config for AM243x EVM to the device tree.

Signed-off-by: Dave Joseph <d-joseph@ti.com>

dnltz reviewed (changes requested):
I think power handling should be handles exactly like pinmuxing, clock and reset; from within the driver.

Which means the driver should have a `power-domains` property similar to Linux. If the driver is enabled, it should automatically enable this node. This would save us all these `i2c0_pd { status = "okay"};` nodes in out device tree.

dnltz commented on dts/arm/ti/am64x_main.dtsi:
```diff
@@ -5,7 +5,7 @@
 
 #include <freq.h>
-
+#include "am64x_power_domains.dtsi"
```
Can we please call this file `am64x_main_...`? The AM64x has a MCU only power mode which turns off the entire main power main but keeps the MCU on. We should separate both to not accidentally manipulate in the wrong power domain.
```suggestion
#include "am64x_main_power_domains.dtsi"

```

dnltz commented on dts/arm/ti/am64x_power_domains.dtsi:
```diff
@@ -0,0 +1,1192 @@
+			zephyr,pm-device-runtime-auto;
+		};
+
+		mcu_m4fss0_cbass_0_pd:mcu_m4fss0_cbass_0_pd {
```
Please move all `mcu_` domains to `am64x_mcu_power_domains`.

dnltz commented on tests/drivers/power/tisci/src/main.c:
```diff
@@ -0,0 +1,48 @@
+	zassert_not_null(pd_dev, "Power domain device not found");
+	zassert_true(device_is_ready(pd_dev), "Power domain device not ready");
+
+	const struct device *dmsc = DEVICE_DT_GET(DT_NODELABEL(dmsc));
```
This node label doesn't exist. You need to rebase after #90053 got merged.
